### PR TITLE
Rewrite OOP section

### LINKED ISSUE #4172: Tiny typo in Waypoint: Make Instances of Objects with a Constructor Function instructions
In the first paragraph, "many copies of an objects" should be "many copies of an object" (object should be singular).

link: http://www.freecodecamp.com/challenges/waypoint-make-instances-of-objects-with-a-constructor-function

screen cap:
fcctypo

### LINKED ISSUE #4177: Waypoint: Make Object Properties Private - Needs Rewrite
Challenge [Waypoint: Make Object Properties Private](http://www.freecodecamp.com/challenges/waypoint-make-object-properties-private) has an issue.

I've noticed that we're getting a lot of questions on this one in the help channels and elsewhere. Vast majority of campers who are asking just don't know what to do at all.  This either needs a massive rewrite or to be broken out into several waypoints.

I think the issue is that there are too many new concepts introduced with no practical explanation.  There is no example code showing `var privateProperty`.

### PR DESCRIPTION
**Pull request by arialaurel**
In an effort to better explain concepts.

Glad to hear any feedback on the new lessons since this is a somewhat significant rewrite of this section.

After this I plan to move onto hopefully adding more lessons to this section, but for now I just wanted to spruce up what we've got.

Closes #4172 and closes #4177 and also tries to address concerns about this section that continue to be expressed in #2984

### CODE REVIEW

File: seed/challenges/object-oriented-and-functional-programming.json [hunk 1]
```diff
@@ -83,103 +89,163 @@
...
         "var myMotorBike = new MotorBike();",
         "",
-        "// Only change code above this line.",
         "",
         "(function() {return JSON.stringify(myMotorBike);})();"
       ],
+      "solutions":[
+        "var Car = function() {\n  this.wheels = 4;\n  this.engines = 1;\n  this.seats = 1;\n};\n\nvar myCar = new Car();\n\nvar MotorBike = function() {\n  this.engines = 1;\n  this.seats = 1;\n  this.wheels = 4;\n};\n\nvar myMotorBike = new MotorBike();\n\n(function() {return JSON.stringify(myMotorBike);})();\n"
+      ],
       "challengeType":1,
```
Comment (BerkeleyTrue) [inline seed/challenges/object-oriented-and-functional-programming.json]: Solutions shouldn't have the last bit of code `\n\n(function() { return JSON.stringify(myMotorBike);})();/n` just the solution themselfs

**Comment by SaintPeter** [comment]
This looks really solid.  I've tested it locally and everything seems to work properly.  I found the instructions to be very clear and easy to follow.

One comment:
On "Waypoint: Make Unique Objects by Passing Parameters to our Constructor"
You might want to add some comments like 
`// Change this constructor`
above the constructor function.  Unlike prior waypoints, this is not part of the example code.

Also, for the line `var myCar` you may add a comment about adding the assignment or something.

Basically, without a bit more direction, I can already hear the questions coming in.  ;)

**Comment by ltegman** [comment]
Thanks for the feedback! Other stuff has come up, but I'm planning to have that implemented and have all challenges in the OOP/Functional section moved over to use the new head/tail feature by the end of the weekend and will update this PR then.

**Comment by ltegman** [comment]
Okay, @SaintPeter, I've done that and made a couple small corrections to the functional lessons. Holding off on implementing the head/tail spec for functional challenges until we get to the root of #4228 and holding off on adding solutions until #4201 is implemented.

**Comment by SaintPeter** [comment]
Waypoint: Make Unique Objects by Passing Parameters to our Constructor
Passing incorrect parameters to the constructor still passes.  We don't check that `myCar` is properly constructed.
`var myCar = new Car(3,4,1);` passes.

Possible test:
`"assert((function(testCar){ return testCar.wheels === 3 && testCar.seats === 1 && testCar.engines === 2;})(myCar), 'message: <code>myCar</code> should have <code>wheels</code> property of <code>3</code>, a <code>seats</code> property of <code>1</code>, and an <code>engines</code> property of <code>2</code>.');",`

-- 

Since your touching this file, it might make sense to move all of the `tail` functions out to the `tail` param . . . if you've got time to do it.  Could always happen later.

File: seed/challenges/object-oriented-and-functional-programming.json [hunk 2]
```diff
@@ -81,105 +88,164 @@
...
+        "var Car = function() {\n  this.wheels = 4;\n  this.engines = 1;\n  this.seats = 1;\n};\n\nvar myCar = new Car();\n\nvar MotorBike = function() {\n  this.engines = 1;\n  this.seats = 1;\n  this.wheels = 4;\n};\n\nvar myMotorBike = new MotorBike();"
+      ],
       "challengeType":1,
       "type": "waypoint"
     },
     {
-      "id":"cf1111c1c15feddfaeb3bdef",
-      "title":"Make Object Properties Private",
+      "id":"cf1111c1c15feddfaeb4bdef",
+      "title":"Make Instances of Objects with a Constructor Function",
```
Comment (BerkeleyTrue) [inline seed/challenges/object-oriented-and-functional-programming.json]: Is this a new challenge or a re-write? If it is, please you MongoDB to create a new ObjectID instead of incrementing the number.

Comment (BerkeleyTrue) [inline seed/challenges/object-oriented-and-functional-programming.json]: NVM, looks like you just re-ordered these challenges.

**Comment by ltegman** [comment]
@SaintPeter Thanks for the attention to detail during this. I've added a check to be sure they passed in three number arguments to the constructor in that lesson. I have moved all the end function to tail for the lessons that don't involve messing with a function, but I am holding off on doing it for the functional programming lessons (and for the arguments in a constructor one) because there is currently a bug in the code editor where if there are no function calls below where a user either types function or alters an existing function then the entire page will lockup, so I've not moved the tail in any challenges where not having one would cause the page to lockup in the course of the user completing the task.

**Comment by SaintPeter** [comment]
@ltegman You may not apprciate my attention to detail when I point out that:
`"assert(typeof (new Car()).nickname === 'undefined', 'message: There should not be a property <code>engines</code> in the object <code>constructor</code>.');",`
`engines` should be `nickname`.

I have gone over the rest of it with a fine-tooth comb and I think if you fix that ONE last thing we can merge it and be done.

Let me also say that I like your work and I hope you'll be amenable to working with me on future curriculum enhancements.

**Comment by ltegman** [comment]
@SaintPeter, hey I'd rather fix it now than get a bug filed later, so I appreciate it! Knowing all you've done for the FCC community your compliments mean a lot! I'd be happy to lend any help I can to future curriculum updates :smile:

**Comment by SaintPeter** [comment]
All looks good!
